File: Rust/73. Set Matrix Zeroes.rs
// Best Solution Using Time: O(mn), Space: O(1)
pub fn set_zeroes(matrix: &mut Vec<Vec<i32>>) {
    let mut row_flag = false;
    let mut col_flag = false;

    for i in 0..matrix.len() {
        for j in 0..matrix[0].len() {
            if matrix[i][j] == 0 {
                matrix[i][0] = 0;
                matrix[0][j] = 0;

                if i == 0 {
                    row_flag = true;
                }
                if j == 0 {
                    col_flag = true;
                }
            }
        }
    }

    for i in 1..matrix.len() {
        for j in 1..matrix[0].len() {
            if matrix[i][0] == 0 || matrix[0][j] == 0 {
                matrix[i][j] = 0;
            }
        }
    }

    if row_flag {
        for j in 0..matrix[0].len() {
            matrix[0][j] = 0;
        }
    }

    if col_flag {
        for i in 0..matrix.len() {
            matrix[i][0] = 0;
        }
    }
}

// Simple Solution using Time: O(mn), Space: O(m + n)
pub fn set_zeroes(matrix: &mut Vec<Vec<i32>>) {
    let mut row: Vec<bool> = Vec::with_capacity(matrix.len());
    let mut col: Vec<bool> = Vec::with_capacity(matrix[0].len());

    for _ in 0..matrix.len() {
        row.push(false);
    }

    for _ in 0..matrix[0].len() {
        col.push(false);
    }

    for i in 0..matrix.len() {
        for j in 0..matrix[0].len() {
            if matrix[i][j] == 0 {
                row[i] = true;
                col[j] = true;
            }
        }
    }

    for i in 0..matrix.len() {
        for j in 0..matrix[0].len() {
            if row[i] || col[j] {
                matrix[i][j] = 0
            }
        }
    }
}

File: Java/73. Set Matrix Zeroes.java

import java.util.HashSet;
import java.util.Set;

class Solution {
    /*
     * Time: O(m*n)
     * Space: O(m+n)
     */
    public void setZeroes_1(int[][] matrix) {
        Set<Integer> uniqueRows = new HashSet<>();
        Set<Integer> uniqueCols = new HashSet<>();

        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                if (matrix[i][j] == 0) {
                    uniqueRows.add(i);
                    uniqueCols.add(j);
                }
            }
        }

        for (Integer row : uniqueRows) {
            for (int j = 0; j < matrix[0].length; j++)
                matrix[row][j] = 0;
        }

        for (Integer col : uniqueCols) {
            for (int i = 0; i < matrix.length; i++)
                matrix[i][col] = 0;
        }
    }

    /*
     * Time: O(m*n)
     * Space: O(1)
     */
    public void setZeroes(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;

        boolean firstRow = false, firstCol = false;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == 0) {
                    matrix[0][j] = 0;
                    matrix[i][0] = 0;

                    if (i == 0)
                        firstRow = true;
                    if (j == 0)
                        firstCol = true;
                }
            }
        }

        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                if (matrix[i][0] == 0 || matrix[0][j] == 0)
                    matrix[i][j] = 0;
            }
        }

        if (firstRow) {
            for (int j = 0; j < n; j++) {
                matrix[0][j] = 0;
            }
        }

        if (firstCol) {
            for (int i = 0; i < m; i++) {
                matrix[i][0] = 0;
            }
        }
    }
}